torchrl/objectives/costs/dreamer.py

```diff
@@ -0,0 +1,295 @@
+        lambda_reco (float): the weight of the reconstruction loss.
+        lambda_reward (float): the weight of the reward loss.
+        reco_loss (Optional[str]): the reconstruction loss.
+        reward_loss (Optional[str]): the reward loss.
+        free_nats (int): the free nats.
+        inversed_free_nats (bool): if True, the free nats are inversed. First we average the kl divergence and then we clamp it to the free nats.
+    """
+
+    def __init__(
+        self,
+        world_model: TensorDictModule,
+        cfg: "DictConfig",
+        lambda_kl: float = 1.0,
+        lambda_reco: float = 1.0,
+        lambda_reward: float = 1.0,
+        reco_loss: Optional[str] = None,
+        reward_loss: Optional[str] = None,
+        free_nats: int = 3,
+        inversed_free_nats: bool = False,
+    ):
+        super().__init__()
+        self.world_model = world_model
+        self.cfg = cfg
+        self.reco_loss = reco_loss if reco_loss is not None else "l2"
+        self.reward_loss = reward_loss if reward_loss is not None else "l2"
+        self.lambda_kl = lambda_kl
+        self.lambda_reco = lambda_reco
+        self.lambda_reward = lambda_reward
+        self.free_nats = free_nats
+        self.inversed_free_nats = inversed_free_nats
+
+    def forward(self, tensordict: TensorDict) -> torch.Tensor:
+        tensordict = tensordict.clone(recurse=False)
+
+        # prepare tensordict: remove time in batch dimensions
+        tensordict.batch_size = tensordict.batch_size[:1]
+        # take the first tensor for prev_posterior_state and prev_belief
+        tensordict["prev_posterior_state"] = torch.zeros_like(
+            tensordict["prev_posterior_state"][:, 0]
+        )
+        tensordict["prev_belief"] = torch.zeros_like(tensordict["prev_belief"][:, 0])
+        tensordict["true_reward"] = tensordict["reward"]
+        del tensordict["reward"]
+
+        tensordict = self.world_model(tensordict)
+        # compute model loss
+        kl_loss = self.kl_loss(
+            tensordict.get("prior_means"),
+            tensordict.get("prior_stds"),
+            tensordict.get("posterior_means"),
+            tensordict.get("posterior_stds"),
+        )
+        reco_loss = (
+            distance_loss(
+                tensordict.get("pixels"),
+                tensordict.get("reco_pixels"),
+                self.reco_loss,
+            )
+            .sum((-1, -2, -3))
+            .mean()
+        )
+        reward_loss = distance_loss(
+            tensordict.get("true_reward"),
+            tensordict.get("reward"),
+            self.reward_loss,
+        ).mean()
+        loss = (
+            self.lambda_kl * kl_loss
+            + self.lambda_reco * reco_loss
+            + self.lambda_reward * reward_loss
+        )
+        return (
+            TensorDict(
+                {
+                    "loss_world_model": loss,
+                    "kl_model_loss": kl_loss,
```

Comment:
can we detatch these values, since one of them is already the aggregation of the three?

the preferred (ie more consistent with other loss modules) implementation IMO would be to return this
```
            TensorDict(
                {
                    "loss_kl": self.lambda_kl * kl_loss,
                    "reco_model_loss": self.lambda_reco * reco_loss,
                    "reward_model_loss": self.lambda_reward * reward_loss,
...
```
and do the sum later on